File: src/main/scala/com/github/nichijou/algorithms/SelectionSort.scala
package com.github.nichijou.algorithms

import com.github.nichijou.utils.RandomArr

/**
  * Created by zk_chs on 16/4/26.
  * 选择排序,运行时间和输入无关
  */
object SelectionSort extends Example{

  override def sort[T <% Ordered[T]](a: Array[T]) = {
    val length = a.length
    for (i <- a.indices){ // 与 0 until length 相同
      var min = i
      for (j <- (i + 1) until length)
        if (less(a(j), a(min))) min = j
      exch(a, i, min)
    }
    show(a)
  }

//  println("SelectionSort".toArray.mkString(" "))
//  sort("SelectionSort".toArray)

  val s = "asmnqoakqwnkdksfpqpowqenvsjhsakewqncammxnckjehuiuqwnejkasjncsjakbdqwijncajxkajdkqjwhejnjkxcnajkfbocviuwetiyu"
  val length = s.length
  println(s.toArray.mkString(" "))
  val start = System.currentTimeMillis()
  sort(s.toArray)
  val end = System.currentTimeMillis()
  println("total: " + (end - start))

//  sort(RandomArr.shuffle(10))

}

File: src/main/java/com/github/algorithms/SelectionSort.java
package com.github.algorithms;

/**
 * Created by zk_chs on 16/4/28.
 */
public class SelectionSort extends Example {

    @Override
    void sort(Comparable[] arr) {
        for (int i = 0; i < arr.length; i++){
            int min = i;
            for (int j = i; j < arr.length; j++)
                if (less(arr[j], arr[min])) min = j;
            exch(arr, i, min);
        }
    }

    public static void main(String[] args) {
        Integer[] arr = RandomArr.shuffle(300000);
//        System.out.println(Arrays.toString(arr));
        SelectionSort sort = new SelectionSort();
        long start = System.nanoTime();
        sort.sort(arr);
        long end = System.nanoTime();
//        System.out.println(Arrays.toString(arr));
        System.out.println("total: " + (end - start));
    }

}
